src/datatypes.jl

```diff
@@ -0,0 +1,406 @@
+################################################################################
+const _LogicalExpr = JuMP.GenericNonlinearExpr{LogicalVariableRef}
+
+"""
+    ConstraintData{C <: JuMP.AbstractConstraint}
+
+A type for storing constraint objects in [`GDPData`](@ref) and any meta-data 
+they possess.
+
+**Fields**
+- `constraint::C`: The constraint.
+- `name::String`: The name of the proposition.
+"""
+mutable struct ConstraintData{C <: JuMP.AbstractConstraint}
+    constraint::C
+    name::String
+end
+
+"""
+    LogicalConstraintIndex
+
+A type for storing the index of a logical constraint.
+
+**Fields**
+- `value::Int64`: The index value.
+"""
+struct LogicalConstraintIndex
+    value::Int64
+end
+
+"""
+    LogicalConstraintRef
+
+A type for looking up logical constraints.
+"""
+struct LogicalConstraintRef
+    model::JuMP.Model # TODO: generalize for AbstractModels
+    index::LogicalConstraintIndex
+end
+
+################################################################################
+#                              DISJUNCT CONSTRAINTS
+################################################################################
+"""
+    DisjunctConstraint
+
+Used as a tag for constraints that will be used in disjunctions. This is done via 
+the following syntax:
+```julia-repl
+julia> @constraint(model, [constr_expr], DisjunctConstraint)
+
+julia> @constraint(model, [constr_expr], DisjunctConstraint(lvref))
+```
+where `lvref` is a [`LogicalVariableRef`](@ref) that will ultimately be associated 
+with the disjunct the constraint is added to. If no `lvref` is given, then one is 
+generated when the disjunction is created.
+"""
+struct DisjunctConstraint
+    indicator::LogicalVariableRef
+end
+
+# Create internal type for temporarily packaging constraints for disjuncts
+struct _DisjunctConstraint{C <: JuMP.AbstractConstraint, L <: LogicalVariableRef}
+    constr::C
+    lvref::L
+end
+
+"""
+    DisjunctConstraintIndex
+
+A type for storing the index of a [`DisjunctConstraint`](@ref).
+
+**Fields**
+- `value::Int64`: The index value.
+"""
+struct DisjunctConstraintIndex
+    value::Int64
+end
+
+"""
+    DisjunctConstraintRef
+
+A type for looking up disjunctive constraints.
+"""
+struct DisjunctConstraintRef
+    model::JuMP.Model # TODO: generalize for AbstractModels
+    index::DisjunctConstraintIndex
+end
+
+################################################################################
+#                              DISJUNCTIONS
+################################################################################
+"""
+    Disjunction <: JuMP.AbstractConstraint
+
+A type for a disjunctive constraint that is comprised of a collection of 
+disjuncts of indicated by a unique [`LogicalVariableRef`](@ref).
+
+**Fields**
+- `disjuncts::Vector{LogicalVariableIndex}`: The disjuncts that comprise the constraint.
+"""
+struct Disjunction <: JuMP.AbstractConstraint
+    disjuncts::Vector{LogicalVariableIndex}
+end
```

Comment:
Should we rename `disjuncts` to `indicators` or `disjunct_indicators` to be more clear?

Reply:
I like indicators